### ISSUE TITLE
Feature request: Option to list available upgrades without performing them

### ISSUE THREAD

**Issue by Phidica** [pipxproject/pipx#149]
I would like to see this as an option to `upgrade` and `upgrade-all`, where in the case of the former it would tell you if an upgrade is available for the given package, and in the latter it would list the packages for which an upgrade is available

**Comment by cs01** [comment]
Maybe an option like `--dry-run`?

**Comment by Phidica** [comment]
Yes, that would work

**Comment by cs01** [comment]
I started implementing this using `pip search PACKAGE`, but it does not do exact matching, nor can it (https://github.com/pypa/pip/issues/3354). If there is an API pipx can use, I am happy to add it. 

Another way to implement this would be to install to a temporary directory, get the version, then erase the temporary directory, but I am going to hold off on that.

**Comment by Phidica** [comment]
Hmm, I didn't realise `pip search` was so fraught with troubles. And it seems they may even consider disabling it in the future (pypa/pip#5216) without a better API powering it

It seems that running `pip install <name>==` will prompt an error message (because a version number is missing after the `==`) containing a list of versions which *are* available. Eg, 
```
$ pip3 install pipx==
Collecting pipx==
  Cache entry deserialization failed, entry ignored
  Could not find a version that satisfies the requirement pipx== (from versions: 0.1.0, 0.12.0.0, 0.12.0.1, 0.12.0.2, 0.12.0.3, 0.12.0.4, 0.12.1.0, 0.12.2.0, 0.12.3.0, 0.12.3.1, 0.12.3.2, 0.12.3.3, 0.13.0.0b1, 0.13.0.0, 0.13.0.1, 0.13.1.0)
No matching distribution found for pipx==
```
Would parsing this output be acceptable for finding the latest available version of a package?

**Comment by Phidica** [comment]
Studying the debug output of `pip`, it appears this is internally implemented by analysing the links on the page `https://pypi.python.org/simple/<name>/` eg https://pypi.python.org/simple/pipx/

**Comment by cs01** [comment]
Looks like this url will work:

https://pypi.org/pypi/PACKAGE/json

**Comment by StaticPH** [comment]
Not sure if it will be of much use to others, but I wrote a proof-of-concept bash function that does the job well enough, if not very elegantly.
I use this on a machine with Bash4.4 and GNU coreutils, but I don't think I used anything that wont also work with the BSD counterparts.
```bash
pipx-outdated() {
	echo "OUTDATED PACKAGES:"
	while read -sr pyPkgName pyPkyVersion; do
		local pypi_latest="$(curl -sS https://pypi.org/simple/${pyPkgName}/ | grep -o '>.*</' | tail -n 1 | grep -o -E '[0-9]([0-9]|[-._])*[0-9]')"
		[ "$pyPkyVersion" != "$pypi_latest" ] && printf "%s\n\tCurrent: %s\tLatest: %s\n" "$pyPkgName" "$pyPkyVersion" "$pypi_latest"
	done < <( pipx list | grep -o 'package.*,' | tr -d ',' | cut -d ' ' -f 2- )
}
```

**Comment by uranusjr** [comment]
The underlying technical necessity here is the same as #464.

**Comment by itsayellow** [comment]
Yes, code that leverages the pip Simple API is probably the way to go for specific searches.  It's a shame the return is in HTML instead of some more data-oriented format, but as mentioned in #464 there are python libraries to help parse the html to usable data.  I'm not sure which of them is the best and/or actively-developed.

**Comment by itsayellow** [comment]
e.g. https://github.com/jwodder/pypi-simple, an interface to the Simple API

**Comment by itsayellow** [comment]
https://discuss.python.org/t/potential-inconsistency-w-pep-503-simple-repo-api/3571/4
The quote:
> That’s correct, pretty much nothing uses /simple/ as far as I’m aware. It’s huge and also doesn’t get purged regularly so it’s usually out of date anyways.

seems a little worrying to me to use the Simple API as a trusted reference.

**Comment by uranusjr** [comment]
The simple API is a trusted source. The comment you’re referencing is taking about the `/simple/` page specifically, which lists all packages on an index, that is not relevant here.

**Comment by itsayellow** [comment]
Oh, I thought they came from the same place.

**Comment by justinmayer** [comment]
Now that [Pipx 0.16 has been released](https://github.com/pipxproject/pipx/releases/tag/0.16.0.0) with the ability to introspect package metadata, any chance someone could take a look at adding a `pipx outdated` command? 🙏

**Comment by J-M0** [comment]
`pip` has the `list --outdated` command. Could you hook into that maybe?

**Comment by justinmayer** [comment]
> `pip` has the `list --outdated` command. Could you hook into that maybe?

A  few days ago I added a (not-yet-published) stanza to my [**`up`** Fish shell plugin](https://github.com/justinmayer/tackle/tree/master/plugins/up) that does exactly that, since `pipx list --outdated` does not exist yet. But hooking into `pip list --outdated` is not as straightforward as it might seem at first glance. For example, the output of `pip list --outdated` includes dependencies and Python packaging infrastructure that is irrelevant to what we actually want to know — that is, whether the package we Pipx-installed is out-of-date. Here's the output for the `pre-commit` virtual environment, which itself is up-to-date but includes output for other out-of-date supporting packages that ostensibly we do not care about:
```
~ ➤ $HOME/.local/pipx/venvs/pre-commit/bin/python -m pip list --outdated --disable-pip-version-check

Package    Version Latest Type
---------- ------- ------ -----
PyYAML     5.3.1   5.4.1  wheel
setuptools 51.1.2  51.3.3 wheel
virtualenv 20.3.1  20.4.0 wheel
```

Ergo, I added a Pipx stanza to my `up outdated` function that does this:

* Pipe `pipx list` output into a variable.
* Extract package names from this list and iterate over them.
* Run `pip list --outdated` for the target package (as shown above).
* If a given line in the output includes the target package name, extract the old and new version numbers.
* Print the package name, old version, and new version to the console.

I think this would be done much better within Pipx itself instead of using Fish shell, but at least I've validated my approach and have a solution that works for me in the interim.

**Comment by itsayellow** [comment]
I've started working on this.

One question, do people have an opinion on how to handle non-PyPI packages (e.g. from git or from a URL)?  Right now it appears that `pip list --outdated` actually attempts to look up such packages on PyPI even if they don't exist there.  It normally just doesn't list them in the outdated list because it can't find the latest version.

My main question is the default behavior for pipx--would the user want to see a list of packages where the "latest version is unknown" or just ignore any git- or URL- based packages for the purposes of `pipx list --outdated`?

**Comment by justinmayer** [comment]
I suggest we ignore any packages for which we can’t reliably indicate related version information.

Regarding the output, my thoughts are that it would look like this:

```
~ ➤ pipx list --outdated
ansible: 2.9.14 < 2.10.5
howdoi: 2.0.7 < 2.0.10
pre-commit: 2.8.1 < 2.9.3
```

**Comment by Phidica** [comment]
I would also prefer ignoring non-PyPI packages. Thanks for the efforts!

**Comment by kkovaletp** [comment]
Is there any progress on this so far?
I have update notifier, which currently checks for updates of my globally installed pip packages by executing `pip3 list --outdated --not-required --format json | jq`
I'd really like to have same functionality for pipx as well

**Comment by gaborbernat** [comment]
PR welcome.

**Comment by chrysle** [comment]
I'd like to take over work on this.

**Comment by inkarkat** [comment]
> Not sure if it will be of much use to others, but I wrote a proof-of-concept bash function[...]]
> 
> ```shell
> pipx-outdated() {
> 	echo "OUTDATED PACKAGES:"
> 	while read -sr pyPkgName pyPkyVersion; do
> 		local pypi_latest="$(curl -sS https://pypi.org/simple/${pyPkgName}/ | grep -o '>.*</' | tail -n 1 | grep -o -E '[0-9]([0-9]|[-._])*[0-9]')"
> 		[ "$pyPkyVersion" != "$pypi_latest" ] && printf "%s\n\tCurrent: %s\tLatest: %s\n" "$pyPkgName" "$pyPkyVersion" "$pypi_latest"
> 	done < <( pipx list | grep -o 'package.*,' | tr -d ',' | cut -d ' ' -f 2- )
> }
> ```

Just taking the last listed package is wrong when dev packages or alpha releases have been published. The [PyPi JSON API](https://warehouse.pypa.io/api-reference/json.html#get--pypi--project_name--json) can be used to query the current version elegantly:
```
curl -sS "https://pypi.org/pypi/${pyPkgName}/json" | jq --raw-output '.info.version'
```

**Comment by StaticPH** [comment]
> Just taking the last listed package is wrong when dev packages or alpha releases have been published. The [PyPi JSON API](https://warehouse.pypa.io/api-reference/json.html#get--pypi--project_name--json) can be used to query the current version elegantly:
> 
> ```
> curl -sS "https://pypi.org/pypi/${pyPkgName}/json" | jq --raw-output '.info.version'
> ```

Of course, it isn't a perfect solution. I had only been sharing a quick and dirty proof-of-concept, but I imagine it's plenty good enough for most people. The version I actually use has evolved a fair bit, and it was only this past June or so that I first found myself wanting to know if there was a newer dev-version of a package, and made `jq` an _optional_ dependency for that feature.
<details>
  <summary>For the curious, here's my current working version of the function, now it's own script file:</summary>

```bash
#! /bin/bash
# List outdated (application-like) packages installed using pipx
__PROGNAME__="${0##*/}"

allowUnstable=0
usage(){
	printf "%s\n" "Usage: $__PROGNAME__ [--unstable]" \
		"       $__PROGNAME__ [-h|--help]"
	printf '    %s\n' \
		'Check all packages installed with pipx for a newer version.' \
		'Without the --unstable flag, only version strings consisting exclusively of numbers and decimal points are considered.' \
		'This means that alpha/beta/dev/unstable/prerelease/etc package versions are ignored by default.' \
		'Note that use of the --unstable flag requires the availability of the command-line JSON processing tool, jq, on the PATH.'
}

if [ "x$1" != 'x' ];then
	case "$1" in
		-h|--help) usage; exit 0;;
		--unstable)
			if ! command -v jq >/dev/null 2>&1; then
				printf 'ERROR: Command not found: "jq".\nThe "%s" flag requires this tool be available on the PATH.\n' "$1"
				exit 1
			fi
			shift;
			allowUnstable=1;;
		*) printf 'ERROR: Unrecognized argument: "%s"\n' "$1"; exit 1;;
	esac
fi

_checkPkg(){
	curl -m 10 -Ss "https://pypi.org/simple/${1}/" | \
		sed -Ee 's@[^>]+>([^<]+)</.+@\1@' -e '/[-.]dev[0-9]*[-.]|[<>]/d' \
			-e 's/.+[-_]([0-9]([0-9]|[-._])*[0-9]).+/\1/' -e '/\./!d' | \
		tail -n 1
}
if [ "$allowUnstable" -eq 1 ]; then
	_checkPkg(){
		curl -m 10 -Ss "https://pypi.org/pypi/${1}/json" | jq --raw-output '.info.version'
	}
fi

_pipx_outdated(){
	let hasOutdated;

	if [ -t 1 ]; then
		# Standard output seems like a file descriptor opened on a terminal
		clearLine(){
			# Using standard xterm-ctrl sequences (which any modern terminal emulator should support most of at this point), clear the current line.

			# Move to character in column 1; normally equivalent to `tput hpa 0`.
			# Support for `tput cha <n>` seems less common, even when the terminal properly responds to the equivalent `\033[nG` sequence.
			printf '\033[1`'
			# Erase all text to the right of the cursor on the current line; normally equivalent to `tput el` for xterm-compatible terminals.
			printf '\033[0K'
		}
		displayTask(){
			clearLine
			printf 'Checking package: "%s"...\r' "$@"
		}
	else
		clearLine(){ :; }
		displayTask(){ :; }
	fi

	while read -sr pyPkgName pyPkyVersion; do
		displayTask "$pyPkgName"
		local pypi_latest="$(_checkPkg "$1")"

		if [ "$pyPkyVersion" != "$pypi_latest" ]; then
			[ ! -v hasOutdated ] && printf "An update is available for the following package(s):\n" && hasOutdated=0
			clearLine
			printf "%s\tCurrent: %s\tLatest: %s\n" "$pyPkgName" "$pyPkyVersion" "$pypi_latest"
		fi
	done < <( pipx list | awk '/package.*,/ { sub(",", "", $3); print $2" "$3; }' )

	if [ "${hasOutdated:-1}" -eq 1 ]; then
		printf "All top-level packages are currently up to date.\n"
	else
		printf "\n\n"
	fi
}

_pipx_outdated
```
</details>

**Comment by powercoconola** [comment]
This would be helpful. Right now if I have 100 packages and I want to automate upgrading them, I have to go through one-by-one with `pipx upgrade`. However, if I had this, I can thread all the calls for all 100 packages and perform checking quicker, then only run `pipx upgrade` on the ones that require updating one-by-one.

**Comment by MindTooth** [comment]
> So the desired command would be `pip list --outdated` or what?

I see that consensus is for `--outdated`, so I would say yes.

For me this would help not to run `pipx upgrade-all` when it's not needed.

Also, having the command return an exit code when updates are available, then we don't need to manipulate the output, but can read the exit status in stead.  E.g. `dnf check-update` returns 100 if updates are available.

**Comment by HeyItsJono** [comment]
Is this a feature that's being considered? Seems to have been 6 years since the initial request but can't see if any developers have acknowledged it.
